rollup/rollup_sync_service/rollup_sync_service.go

```diff
@@ -225,22 +225,37 @@ func (s *RollupSyncService) parseAndUpdateRollupEventLogs(logs []types.Log, endB
 			batchIndex := event.BatchIndex.Uint64()
 			log.Trace("found new FinalizeBatch event", "batch index", batchIndex)
 
-			parentBatchMeta, chunks, err := s.getLocalInfoForBatch(batchIndex)
-			if err != nil {
-				return fmt.Errorf("failed to get local node info, batch index: %v, err: %w", batchIndex, err)
+			lastFinalizedBatchIndex := rawdb.ReadLastFinalizedBatchIndex(s.db)
+			// When failed to get lastFinalizedBatchIndex, set it's value to previous one of current batch index.
+			// It's backward compatible. (works for both before and after darwin)
+			if lastFinalizedBatchIndex == nil {
+				log.Trace("got nil when reading last finalized batch index.")
+				previousBatchIndex := batchIndex - 1
+				lastFinalizedBatchIndex = &previousBatchIndex
 			}
 
-			endBlock, finalizedBatchMeta, err := validateBatch(event, parentBatchMeta, chunks, s.bc.Config(), s.stack)
-			if err != nil {
-				return fmt.Errorf("fatal: validateBatch failed: finalize event: %v, err: %w", event, err)
-			}
+			for index := *lastFinalizedBatchIndex + 1; index <= batchIndex; index++ {
+				parentBatchMeta, chunks, err := s.getLocalInfoForBatch(index)
+				if err != nil {
+					return fmt.Errorf("failed to get local node info, batch index: %v, err: %w", index, err)
+				}
 
-			rawdb.WriteFinalizedL2BlockNumber(s.db, endBlock)
-			rawdb.WriteFinalizedBatchMeta(s.db, batchIndex, finalizedBatchMeta)
+				endBlock, finalizedBatchMeta, err := validateBatch(event, parentBatchMeta, chunks, s.bc.Config(), s.stack)
+				if err != nil {
+					return fmt.Errorf("fatal: validateBatch failed: finalize event: %v, err: %w", event, err)
+				}
+
+				if index == batchIndex {
+					rawdb.WriteFinalizedL2BlockNumber(s.db, endBlock)
```

Comment:
can put this and `rawdb.WriteLastFinalizedBatchIndex(s.db, batchIndex)` together.

Reply:
nice catch

Reply:
job is done